Runtime/Scripts/BlendShapesController.cs

```diff
@@ -47,107 +52,130 @@ public class BlendShapesController : MonoBehaviour, IUsesStreamReader
+        }
 
-            var streamSettings = streamSource.streamSettings;
-            if (streamSettings == null)
-            {
-                Debug.LogError("Disabling BlendShapesController. No stream settings", this);
-                enabled = false;
-                return;
-            }
+        void OnDisable()
+        {
+            UnityEditor.AssemblyReloadEvents.afterAssemblyReload -= OnAssemblyReload;
+        }
 
-            var blendShapesCount = streamSettings.BlendShapeCount;
-            if (m_Overrides.Length != blendShapesCount)
-                Array.Resize(ref m_Overrides, blendShapesCount);
+        void OnAssemblyReload()
+        {
+            UpdateBlendShapeIndices();
+        }
+#endif
 
-            var filteredList = new List<SkinnedMeshRenderer>();
-            foreach (var renderer in m_SkinnedMeshRenderers)
-            {
-                if (renderer == null)
-                {
-                    Debug.LogWarning("Null element in SkinnedMeshRenderer list in " + this);
-                    continue;
-                }
+        void Awake()
+        {
+            m_SmoothedBlendShapes = new Vector2[m_BlendShapeValues.Count];
 
-                if (renderer.sharedMesh == null)
-                {
-                    Debug.LogWarning("Missing mesh in " + renderer);
-                    continue;
-                }
+            UpdateBlendShapeIndices();
 
-                filteredList.Add(renderer);
-            }
+            if (m_Overrides.Length != m_BlendShapeValues.Count)
+                Array.Resize(ref m_Overrides, m_BlendShapeValues.Count);
+
+#if UNITY_EDITOR
+            m_LastTime = Time.realtimeSinceStartup;
+#endif
+        }
 
-            m_SkinnedMeshRenderers = filteredList.ToArray();
+        void LateUpdate()
+        {
+            var deltaTime = Time.deltaTime;
 
-            if (m_SkinnedMeshRenderers.Length < 1)
+#if UNITY_EDITOR
+            if (!Application.isPlaying)
             {
-                Debug.LogWarning("Blend Shape Controller has no valid Skinned Mesh Renderers.");
-                enabled = false;
+                var currentTime = Time.realtimeSinceStartup;
+                deltaTime = currentTime - m_LastTime;
+                m_LastTime = currentTime;
             }
+#endif
+
+            UpdateBlendShapes(deltaTime);
+        }
+
+        public void UpdateBlendShapes(float deltaTime)
+        {
+            UpdateFromStreamReader();
+            PostProcessValues(deltaTime);
+            UpdateSkinnedMeshRenderers();
         }
 
-        void Update()
+        void UpdateFromStreamReader()
         {
+            if (streamReader == null)
+                return;
+            
             var streamSource = streamReader.streamSource;
             if (streamSource == null || !streamSource.isActive)
                 return;
+            
+            isTrackingActive = !streamReader.faceTrackingLost;
```

Comment:
I would prefer to just read the streamReader value directly rather than introducing a new variable unless there is a good reason. Thoughts?

Reply:
To make BlendShapeController work in EditMode using baked AnimationClips, we need a new serialized variable to allow the animation system to create a curve binding.
In this case, we set the value from the stream reader if present.

Reply:
Understood